STM32F1/cores/maple/libmaple/i2c.c

```diff
@@ -409,6 +409,24 @@ int32 i2c_master_xfer(i2c_dev *dev,
         // with state changes in the IRQ handlers:
         dev->state = I2C_STATE_ERROR;
         if (rc == I2C_ERROR_TIMEOUT) dev->error_flags |= I2C_SR1_TIMEOUT;
+        if (!(dev->config_flags & I2C_SLAVE_MODE) &&
+            (dev->error_flags & (I2C_SR1_AF | I2C_SR1_BERR | I2C_SR1_TIMEOUT))
+            ) {    // In Master Mode, we need to abort the transmission with a STOP
+                   // for NACK, Bus Error, or Timeout
+            uint32 cr1;
+            while ((cr1 = dev->regs->CR1) & (I2C_CR1_START |
+                                             I2C_CR1_STOP  |
+                                             I2C_CR1_PEC)) {
+                ;   // Must wait for pending start/stop/pec before setting stop to avoid
+                    //  accidental restart condition. See ST RM0008 Note in 26.6.1 (I2C_CR1)
+            }
+            dev->regs->CR1 |= I2C_CR1_STOP;
+            while ((cr1 = dev->regs->CR1) & (I2C_CR1_START |
+                                             I2C_CR1_STOP  |
+                                             I2C_CR1_PEC)) {
+                ;
```

Comment:
Why to check for START and PEC since only START was set from the last check? START and PEC flags are for sure not set at this point.

Reply:
From ST's Reference Manual 26.6.1 that I referenced:

> Note: When the STOP, START or PEC bit is set, the software must not perform any write access to I2C_CR1 before this bit is cleared by hardware. Otherwise there is a risk of setting a second STOP, START or PEC request.

Also, if you go back and look at the original `i2c.c` code (I think it was in the `i2c_common.h` file) that's exactly what they had in the setting of the STOP.  I did away with that macro that did that because 1) you don't want to do that loop in many of the cases of setting STOP -- for example, you absolutely don't want such a `while` loop in an IRQ handler as you'll starve the system and/or deadlock.  2) it was only needed in this one location.

Essentially ANYTIME you write CR1, you should make sure that all 3 flags are off.  It's actually not necessarily true that the START is no longer set here if there was a bus error or something in the process of setting the START in the first place -- you must remember that this isn't just for the NACK, but any error, including bus error and timeout.

Reply:
Well, I am not quite convinced, but anyway, the code size to check one bit or more is the same so I accept that.

Something else: why is necessary to use a local variable `cr1`? The register accesses are per definition volatile.

Reply:
The local variable shouldn't be needed and will be optimized out by the compiler.  This code was a cut-and-paste from the `i2c.h` code from way back long before I had made any changes at all to I2C and had my paws in it (and I figured you would be more accepting of the code if it was a bit more like it originally was).  Here's that code from the commit that was at the head of master when I originally started working on I2C way back when: https://github.com/rogerclarkmelbourne/Arduino_STM32/blob/60d982ff25d966f15e8af70d4bb6f46bb1d1595a/STM32F1/system/libmaple/include/libmaple/i2c.h#L242-L256

Reply:
It may be but I still consider that it is not needed. I tried without `cr1` and it works well.
Also, the second check after setting the STOP bit is also not needed, the transfer function has anyway at the beginning a check for I2C busy, so the application will wait there if the HW is still busy. Meanwhile the application has more time to do other stuff without waiting the STOP to end.

Reply:
Actually, I disagree about the second wait for the STOP.  Now in practice, this F1 CPU is probably slow enough that you can get away with it (by "accident"), but if there are any writes to the CR1 register before the peripheral has processed that STOP and set the BUSY busy flag (such as someone immediately calling I2C again to send another message), then it's possible that the read-modify-write operation of the CR1 setting would end up setting it again.   The code should be written so that it doesn't assume that the processor is slow enough to get it processed before we do anything else.

The amount of time that while loop 'wastes' from the application is very minimal and interrupts are still enabled and can be processed, so other peripherals aren't delayed at all.  I would personally be surprised if it makes more than one pass through the loop.  But it can potentially prevent a race-condition on a CR1 write.

Reply:
Usually, CR1 will be written eventually by the next transfer function. So the best place to check for busy and/or START/STOP/PEC is at the beginning of the function. The busy flag is already checked there, it cold be extended with those 3 additional flags, although I think the busy flag is anyway set during start/stop/pec.
I measured ~10 µs till the second check will is finished, this time could be used by the application for other things. Not that it is much, but in principle I think this is the way how it should be done. And the START bit also not checked till is finished, the handling would then be identical for both flags.

Reply:
Well, first off, that loop is only run in the error condition, not every time a stop is sent on the busy.  If the I2C is in such an error state, I doubt it will notice the 10µSec delay (and if you are scanning the bus or something, you have to wait that time period anyway!).  And there's no need to check either START or STOP under normal conditions as the IRQ handles that and the flags will be clear before transitioning to the next message.  That's why I removed that `i2c_stop_condition` function in the original code as there shouldn't be a such a `while` loop on every stop.  In fact, if that function were called in an IRQ, either the loop is a no-op or you hang indefinitely.  The only place it makes sense is this error condition where you are explicitly triggering a `STOP` when the slave fails to respond.  But, without the second `while` loop, there is a potential race-condition between that `STOP` and the start of another message.

Reply:
I do not think that there will be a race condition. I tested the scanner without the second while loop and it works correctly. Meaning, the START will not interfere with the STOP.
The passage from the reference manual says (as you cited):
> Note: When the STOP, START or PEC bit is set, the software must not perform any write access to I2C_CR1 before this bit is cleared by hardware. Otherwise there is a risk of setting a second STOP, START or PEC request.

It means that if one bit is set than you eventually should wait for that flag to be cleared in order to avoid a second write of the same bit, not of other bit. And as I told, setting START without waiting for STOP to end works correctly. And you also don't wait for START to end, right?

Reply:
The Read-Modify-Write operation causes the problem.  If there's one of these three bits still set, then the process of reading CR1 to `OR` in the bit you are setting (be it `START` or `STOP`) then that other bit will still be set when you then write it back to CR1.  That write then triggers the I2C peripheral to do yet another `START`, etc., operation.  On this I2C peripheral, the writing of a `1` to any of those bits triggers that operation.  So, as the manual says, you must make sure that those three flags are all clear before doing any such write operation to the CR1 register.  This is what can trigger the restart condition like you were seeing before, for example.  If `START` was still set from the previous attempt to write to a nonexistent slave, then the act of writing the `STOP` could trigger a restart.

Now, in all due fairness, the second `WHILE` loop could be modified to just include the `STOP`, as the other two flags cannot be set there like they potentially could be for the first `WHILE` loop, since you only ever wrote a `STOP` in between the two `WHILE` loops ... but to keep the code consistent for readability purposes, leaving them the same makes sense to do.  The compiler will combine the three bits into a single value anyway, so execution-wise it doesn't matter if you check one bit or all three.  It's a moot point.

Reply:
I understand the reason for the first loop, what i've never contested. I was always talking about the second while loop, in particular the reason for it.
In my opinion, identical to what we have here (the first while loop), that should be inserted before writing of START and PEC bits instead of having the second while loop (waiting for one bit to be cleared by the hardware right after setting it). The second while loop causes unnecessary dead time.
Again, there is no second while loop after setting the START bit, right?

Reply:
Right -- there is no while loop after setting the START bit because no one else is writing to CR1 until the message is complete.  But, here when writing the STOP bit, the app (which now thinks that I2C processing is complete) could immediately call another function that could potentially write CR1 -- and NOT just the case of starting a new message, but anything, like reconfiguring I2C.

Reply:
That was a convincing argument. I do not know how often would be the case to reconfigure the I2C after a STOP, but this change would anyway keep things safe, despite introducing a small delay.
I have tested it, and I am going to merge this.